File: packages/chat-python/src/rfnry_chat_protocol/event.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter

from rfnry_chat_protocol.content import ContentPart
from rfnry_chat_protocol.identity import Identity
from rfnry_chat_protocol.tenant import TenantScope


class _EventBase(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    id: str
    thread_id: str
    run_id: str | None = None
    author: Identity
    created_at: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)
    client_id: str | None = None
    recipients: list[str] | None = None


class ToolCall(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    name: str
    arguments: Any


class ToolResult(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    result: Any | None = None
    error: dict[str, Any] | None = None


class _EventError(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    code: str
    message: str


class _ThreadCreatedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    tenant: TenantScope


class MessageEvent(_EventBase):
    type: Literal["message"] = "message"
    content: list[ContentPart]


class ReasoningEvent(_EventBase):
    type: Literal["reasoning"] = "reasoning"
    content: str


class ToolCallEvent(_EventBase):
    type: Literal["tool.call"] = "tool.call"
    tool: ToolCall


class ToolResultEvent(_EventBase):
    type: Literal["tool.result"] = "tool.result"
    tool: ToolResult


class ThreadCreatedEvent(_EventBase):
    type: Literal["thread.created"] = "thread.created"
    thread: _ThreadCreatedPayload


class ThreadMemberAddedEvent(_EventBase):
    type: Literal["thread.member_added"] = "thread.member_added"
    member: Identity


class ThreadMemberRemovedEvent(_EventBase):
    type: Literal["thread.member_removed"] = "thread.member_removed"
    member: Identity


class ThreadTenantChangedEvent(_EventBase):
    type: Literal["thread.tenant_changed"] = "thread.tenant_changed"
    from_: TenantScope = Field(alias="from", serialization_alias="from")
    to: TenantScope


class RunStartedEvent(_EventBase):
    type: Literal["run.started"] = "run.started"


class RunCompletedEvent(_EventBase):
    type: Literal["run.completed"] = "run.completed"


class RunFailedEvent(_EventBase):
    type: Literal["run.failed"] = "run.failed"
    error: _EventError


class RunCancelledEvent(_EventBase):
    type: Literal["run.cancelled"] = "run.cancelled"


Event = (
    MessageEvent
    | ReasoningEvent
    | ToolCallEvent
    | ToolResultEvent
    | ThreadCreatedEvent
    | ThreadMemberAddedEvent
    | ThreadMemberRemovedEvent
    | ThreadTenantChangedEvent
    | RunStartedEvent
    | RunCompletedEvent
    | RunFailedEvent
    | RunCancelledEvent
)

_event_adapter: TypeAdapter[Event] = TypeAdapter(
    Annotated[Event, Field(discriminator="type")],
)


def parse_event(raw: dict[str, Any]) -> Event:
    return _event_adapter.validate_python(raw)


class EventDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    client_id: str
    content: list[ContentPart] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    recipients: list[str] | None = None

File: packages/chat-typescript/src/event.ts
import type { ContentPart, ContentPartWire } from './content'
import type { Identity, IdentityWire } from './identity'
import type { TenantScope } from './tenant'

export type EventBase = {
  id: string
  threadId: string
  runId?: string
  author: Identity
  createdAt: string
  metadata: Record<string, unknown>
  clientId?: string
  recipients: string[] | null
}

export type EventBaseWire = {
  id: string
  thread_id: string
  run_id?: string | null
  author: IdentityWire
  created_at: string
  metadata: Record<string, unknown>
  client_id?: string | null
  recipients?: string[] | null
}

export type ToolCall = {
  id: string
  name: string
  arguments: unknown
}

export type ToolResult = {
  id: string
  result?: unknown
  error?: { code: string; message: string }
}

export type MessageEvent = EventBase & { type: 'message'; content: ContentPart[] }
export type ReasoningEvent = EventBase & { type: 'reasoning'; content: string }
export type ToolCallEvent = EventBase & { type: 'tool.call'; tool: ToolCall }
export type ToolResultEvent = EventBase & { type: 'tool.result'; tool: ToolResult }
export type ThreadCreatedEvent = EventBase & {
  type: 'thread.created'
  thread: { id: string; tenant: TenantScope }
}
export type ThreadMemberAddedEvent = EventBase & {
  type: 'thread.member_added'
  member: Identity
}
export type ThreadMemberRemovedEvent = EventBase & {
  type: 'thread.member_removed'
  member: Identity
}
export type ThreadTenantChangedEvent = EventBase & {
  type: 'thread.tenant_changed'
  from: TenantScope
  to: TenantScope
}
export type RunStartedEvent = EventBase & { type: 'run.started' }
export type RunCompletedEvent = EventBase & { type: 'run.completed' }
export type RunFailedEvent = EventBase & {
  type: 'run.failed'
  error: { code: string; message: string }
}
export type RunCancelledEvent = EventBase & { type: 'run.cancelled' }

export type Event =
  | MessageEvent
  | ReasoningEvent
  | ToolCallEvent
  | ToolResultEvent
  | ThreadCreatedEvent
  | ThreadMemberAddedEvent
  | ThreadMemberRemovedEvent
  | ThreadTenantChangedEvent
  | RunStartedEvent
  | RunCompletedEvent
  | RunFailedEvent
  | RunCancelledEvent

export type EventWire = EventBaseWire & {
  type: Event['type']
} & Record<string, unknown>

export type EventDraft = {
  clientId: string
  content?: ContentPart[]
  metadata?: Record<string, unknown>
  recipients?: string[] | null
}

export type EventDraftWire = {
  client_id: string
  content?: ContentPartWire[]
  metadata?: Record<string, unknown>
  recipients?: string[] | null
}

const EVENT_TYPES: ReadonlySet<Event['type']> = new Set([
  'message',
  'reasoning',
  'tool.call',
  'tool.result',
  'thread.created',
  'thread.member_added',
  'thread.member_removed',
  'thread.tenant_changed',
  'run.started',
  'run.completed',
  'run.failed',
  'run.cancelled',
])

const IDENTITY_ROLES = new Set(['user', 'assistant', 'system'])

function toIdentityInternal(wire: IdentityWire): Identity {
  return {
    role: wire.role,
    id: wire.id,
    name: wire.name,
    metadata: wire.metadata ?? {},
  } as Identity
}

function toContentPartInternal(wire: ContentPartWire): ContentPart {
  switch (wire.type) {
    case 'text':
      return wire
    case 'image':
      return wire
    case 'audio':
      return {
        type: 'audio',
        url: wire.url,
        mime: wire.mime,
        name: wire.name,
        size: wire.size,
        durationMs: wire.duration_ms,
      }
    case 'document':
      return wire
    case 'form':
      return {
        type: 'form',
        formId: wire.form_id,
        schema: wire.schema,
        status: wire.status,
        values: wire.values,
        answersEventId: wire.answers_event_id,
        title: wire.title,
        description: wire.description,
      }
  }
}

export function parseEvent(raw: unknown): Event {
  if (typeof raw !== 'object' || raw === null) {
    throw new Error(`invalid event: ${JSON.stringify(raw)}`)
  }
  const record = raw as Record<string, unknown>
  const type = record.type
  if (typeof type !== 'string' || !EVENT_TYPES.has(type as Event['type'])) {
    throw new Error(`unknown event type: ${JSON.stringify(type)}`)
  }
  if (typeof record.id !== 'string' || typeof record.thread_id !== 'string') {
    throw new Error(`invalid event base: ${JSON.stringify(raw)}`)
  }
  if (typeof record.created_at !== 'string') {
    throw new Error(`invalid event created_at: ${JSON.stringify(raw)}`)
  }
  const author = record.author as IdentityWire | undefined
  if (
    !author ||
    typeof author.id !== 'string' ||
    typeof author.role !== 'string' ||
    !IDENTITY_ROLES.has(author.role)
  ) {
    throw new Error(`invalid event author: ${JSON.stringify(raw)}`)
  }
  const base: EventBase = {
    id: record.id,
    threadId: record.thread_id,
    runId: (record.run_id as string | null | undefined) ?? undefined,
    author: toIdentityInternal(author),
    createdAt: record.created_at,
    metadata: (record.metadata as Record<string, unknown> | undefined) ?? {},
    clientId: (record.client_id as string | null | undefined) ?? undefined,
    recipients: (record.recipients as string[] | null | undefined) ?? null,
  }
  switch (type as Event['type']) {
    case 'message':
      return {
        ...base,
        type: 'message',
        content: ((record.content as ContentPartWire[] | undefined) ?? []).map(
          toContentPartInternal
        ),
      }
    case 'reasoning':
      return { ...base, type: 'reasoning', content: (record.content as string) ?? '' }
    case 'tool.call':
      return { ...base, type: 'tool.call', tool: record.tool as ToolCall }
    case 'tool.result':
      return { ...base, type: 'tool.result', tool: record.tool as ToolResult }
    case 'thread.created':
      return {
        ...base,
        type: 'thread.created',
        thread: record.thread as { id: string; tenant: TenantScope },
      }
    case 'thread.member_added':
      return {
        ...base,
        type: 'thread.member_added',
        member: toIdentityInternal(record.member as IdentityWire),
      }
    case 'thread.member_removed':
      return {
        ...base,
        type: 'thread.member_removed',
        member: toIdentityInternal(record.member as IdentityWire),
      }
    case 'thread.tenant_changed':
      return {
        ...base,
        type: 'thread.tenant_changed',
        from: record.from as TenantScope,
        to: record.to as TenantScope,
      }
    case 'run.started':
      return { ...base, type: 'run.started' }
    case 'run.completed':
      return { ...base, type: 'run.completed' }
    case 'run.failed':
      return {
        ...base,
        type: 'run.failed',
        error: record.error as { code: string; message: string },
      }
    case 'run.cancelled':
      return { ...base, type: 'run.cancelled' }
  }
}
